File: validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* prev = nullptr; 
     bool isValidBST(TreeNode* root) {
        
        if(root == nullptr)
            return true; 
        
        if(isValidBST(root->left) == false)
            return false; 
        
        
        if(prev != nullptr &&  prev->val >= root->val)
            return false; 
        
        prev = root; 
        
        if(isValidBST(root->right) == false)
            return false; 
        
        return true; 
   
    }
};

File: validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

 List<Integer> str = new LinkedList<>(); 
  public void Inorder(TreeNode root)
    {
        if(root == null) //BASE CASE 
            return; 

        //LMR
        Inorder(root.left);//L
        str.add(root.val); //M
        Inorder(root.right);//R
    }
    
    public boolean isValidBST(TreeNode root) {
        Inorder(root); 
        for(int i =0;i<str.size()- 1;i++){
            System.out.println("here"); 
            if(str.get(i) >= str.get(i + 1)){
                return false;
            }
        }
        return true; 
    }

}